File: solutions-neetcode/python/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(root):
            nonlocal res

            if not root:
                return 0
            left = dfs(root.left)
            right = dfs(root.right)
            res = max(res, left + right)

            return 1 + max(left, right)

        dfs(root)
        return res





# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = 0
        
        #returns the height for the current node
        def height(root):
            nonlocal diameter
            
            if not root:
                return 0
            
            #returns the max height on the left side
            left = height(root.left)
            #returns the max height on the right side
            right = height(root.right)

            #updates the max diameter by adding both sides
            diameter = max(diameter, left + right)

            curr_height = max(left, right) + 1

            return curr_height
        
        height(root)

        return diameter

File: solutions-neetcode/javascript/0543-diameter-of-binary-tree.js
/**
 * https://leetcode.com/problems/diameter-of-binary-tree/
 * TIme O(N) | Space O(H)
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root, max = [0]) {
    diameterOfTree(root, max);

    return max[0];
};

const diameterOfTree = (root, max) => {
    const isBaseCase = root === null;
    if (isBaseCase) return 0;

    return dfs(root, max);
}

const dfs = (root, max) => {
    const left = diameterOfTree(root.left, max);
    const right = diameterOfTree(root.right, max);

    const diameter = left + right;
    max[0] = Math.max(max[0], diameter);

    const height = Math.max(left, right);

    return height + 1;
}

// claudia
// t o(n) - s o(n) 
var diameterOfBinaryTree = function(root) {
    let maxD = 0

    const dfs = (node) => {
        if (!node) return 0
        const left = dfs(node.left)
        const right = dfs(node.right)
        const currD = left + right
        maxD = Math.max(maxD, currD)
        return Math.max(left, right) + 1
    }
    dfs(root)

    return maxD
};